synfig-studio/src/gui/states/state_bone.cpp

```diff
@@ -1036,108 +904,71 @@ StateBone_Context::event_layer_selection_changed_handler(const Smach::event& /*x
 }
 
 int
-StateBone_Context::find_bone(Point point,Layer::Handle layer,int lay)const
+StateBone_Context::find_bone(Point point,Layer::Handle layer) const
 {
-	if(lay==0){
-		layer = Layer_Skeleton::Handle::cast_dynamic(layer);
-		ValueDesc list_desc(layer,"bones");
-		std::vector<Bone> list=ValueNode_StaticList::Handle::cast_dynamic(list_desc.get_value_node())->operator()(get_canvas()->get_time()).get_list_of(Bone());
-		Real close_line(10000000),close_origin(10000000);
-		Vector direction;
-		Angle angle;
-		int ret;
-		Matrix m;
-		for(auto iter=list.begin();iter!=list.end();++iter){
-
-			m=iter->get_animated_matrix();
-			Point orig = m.get_transformed(Vector(0,0));
-			angle = Angle::rad(atan2(m.axis(0)[1],m.axis(0)[0]));
-			Real orig_dist((point-orig).mag());
-			Real dist=std::fabs(orig_dist*Angle::sin((point-orig).angle()-angle).get());
-			Real length = iter->get_length()*iter->get_scalelx();
-			if(Angle::cos((point-orig).angle()-angle).get()>0 && orig_dist<=length){
-				dist = std::fabs(dist);
-				if(dist<close_line){
-					close_line=dist;
-					close_origin=orig_dist;
-					ret = iter-list.begin();
-				}else if(fabs(dist-close_line)<0.0000001 && close_line!= 10000000){
-					if(orig_dist<close_origin){
-						close_origin=orig_dist;
-						ret = iter-list.begin();
-					}
-				}
-			}
-
-		}
-		if(std::fabs(close_line)<=0.2){
-			if (ret < static_cast<int>(list.size()))
-				return ret;
-			else
-				return -1;
-		}else{
-			return -1;
-		}
+	std::vector<Bone> bone_list;
+	ValueDesc list_desc(layer,"bones");
+	if(Layer_Skeleton::Handle::cast_dynamic(layer)){
+		bone_list = (*list_desc.get_value_node())(get_canvas()->get_time()).get_list_of(Bone());
+	} else if (Layer_SkeletonDeformation::Handle::cast_dynamic(layer)) {
+		std::vector<std::pair<Bone,Bone>> bone_pair_list = (*list_desc.get_value_node())(get_canvas()->get_time()).get_list_of(std::pair<Bone,Bone>());
+		for (const auto& item : bone_pair_list)
+			bone_list.push_back(item.second);
```

Comment:
After reviewing this part I have realized that there is a bug in original code: it should use "first" bone instead of "second". Because of this bug I was experiencing an issue when it was not possible to select bone as active at some when second bone had different position than second (this usually happens when user deforms image).

So, we should change this part of code from "item.second" to "item.first', but I suggest to do that in separate PR, because it's better to have this PR for cleanup&refactoring only. ^__^

Reply:
First item is for resting and second is for posing state, right?

Reply:
Maybe it depends on layer activate status?

Reply:
> First item is for resting state and the second one is for posing state, right?

Exactly!

Reply:
> Maybe it depends on layer activate status?

In general case - yes. But Bone Tool always sets the Skeleton Deformation layer to “inactive” state. So, in our case we always deal with “first”.

Reply:
Fixed via https://github.com/synfig/synfig/pull/2340

Reply:
Thank you! ^__^